byron-proxy/src/exec/Main.hs

```diff
@@ -401,15 +430,90 @@ runClient tracer clientOptions genesisConfig epochSlots db = case clientOptions
   Nothing -> pure ()
 
   Just (Left shelleyClientOptions) -> do
-    addrInfos <- Network.getAddrInfo (Just addrInfoHints) (Just host) (Just port)
-    case addrInfos of
-      [] -> error "no getAddrInfo"
-      (addrInfo : _) -> IPC.runClient addrInfo stringTracer epochSlots db
+    addrInfosLocal  <- Network.getAddrInfo (Just addrInfoHints) (Just "127.0.0.1") (Just "0")
+    addrInfosRemote <- Network.getAddrInfo (Just addrInfoHints) (Just host) (Just port)
+    case (addrInfosLocal, addrInfosRemote) of
+      (addrInfoLocal : _, addrInfoRemote : _) -> connectTo
+        encodeTerm
+        decodeTerm
+        (initiatorVersions epochSlots chainSyncClient)
+        addrInfoLocal
+        addrInfoRemote
+      _ -> error "no getAddrInfo"
     where
 
     host = scoHostName shelleyClientOptions
     port = scoServiceName shelleyClientOptions
     addrInfoHints = Network.defaultHints
+    -- | This chain sync client will first try to improve the read pointer to
+    -- the tip of the database, and then will roll forward forever, stopping
+    -- if there is a roll-back.
+    -- It makes sense given that we only have an immutable database and one
+    -- source for blocks: one read pointer improve is always enough.
+    chainSyncClient = Client.chainSyncClient fold
+      where
+      fold :: Client.Fold (ResourceT IO) x
+      fold = Client.Fold $ do
+        tip <- lift $ DB.readTip db
+        mPoint <- case tip of
+          -- DB is empty. Can go without improving read pointer.
+          DB.TipGenesis -> pure Nothing
+          -- EBB is nice because we already have the header hash.
+          DB.TipEBB   slotNo hhash _     -> pure $ Just $ ChainSync.Point
+            { ChainSync.pointSlot = slotNo
+            , ChainSync.pointHash = hhash
+            }
+          DB.TipBlock slotNo       bytes -> pure $ Just $ ChainSync.Point
+            { ChainSync.pointSlot = slotNo
+            , ChainSync.pointHash = hhash
+            }
+            where
+            hhash = case Binary.decodeFullAnnotatedBytes "Block or boundary" (Cardano.fromCBORABlockOrBoundary epochSlots) bytes of
+              Left cborError -> error "failed to decode block"
```

Comment:
Will this error message be enough to easily find where the error happen? I usually include function name or some other identifier.

Reply:
When this happens, the database is corrupt, so no matter where it came from, the source of the error is the same.

I've been meaning to put this error into the `DB` abstraction instead. What do you think about that? It would mean the `DB` would have to run a decoder for every read, but would make it easier to use.

Reply:
I am not familiar with the `DB` abstraction you have, if you think it will make it easier to use then that might be a good idea. Laziness will help to save work if one does not need parsing.

Reply:
Laziness wouldn't necessarily be of much help. Most of the work is presumably in the decoder, which of course would have to run to completion.